File: app/src/main/java/com/example/unsplash/view/adapters/MyPagedListAdapter.java
package com.example.unsplash.view.adapters;

import android.arch.paging.PagedListAdapter;
import android.content.Context;
import android.support.annotation.NonNull;
import android.support.v7.util.DiffUtil;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import com.example.unsplash.R;
import com.example.unsplash.model.models.Photo;
import com.squareup.picasso.Picasso;

public class MyPagedListAdapter extends PagedListAdapter<Photo, MyPagedListAdapter.MyPagedViewHolder> {

    private Context mContext;
    private PagedListOnClickListener mListener;

    public MyPagedListAdapter(Context context, PagedListOnClickListener listener) {
        super(DIFF_CALLBACK);
        this.mContext = context;
        this.mListener = listener;
    }

    class MyPagedViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        ImageView image;
        PagedListOnClickListener mpListener;

        MyPagedViewHolder(View view, PagedListOnClickListener listener) {
            super(view);
            image = view.findViewById(R.id.picture);
            mpListener = listener;
            view.setOnClickListener(this);
        }

        @Override
        public void onClick(View v) {
            mpListener.onClick(image, getItem(getAdapterPosition()));
        }
    }

    @NonNull
    @Override
    public MyPagedListAdapter.MyPagedViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
        View view;
        view = LayoutInflater.from(mContext).inflate(R.layout.item_picture_main_activity, viewGroup, false);
        return new MyPagedViewHolder(view, mListener);
    }

    @Override
    public void onBindViewHolder(@NonNull final MyPagedListAdapter.MyPagedViewHolder myPagedViewHolder, final int i) {
        final Photo photo = getItem(i);
        myPagedViewHolder.image.setTransitionName("" + i);
        if (photo != null) {
            Picasso.get().load(photo.getUrls().getRegular()).noFade().into(myPagedViewHolder.image);
        } else {
            Log.d("mLog", "onBindViewHolder: photo is null");
        }
    }

    private static DiffUtil.ItemCallback<Photo> DIFF_CALLBACK = new DiffUtil.ItemCallback<Photo>() {
        @Override
        public boolean areItemsTheSame(@NonNull Photo photo, @NonNull Photo t1) {
            return photo.equals(t1);
        }

        @Override
        public boolean areContentsTheSame(@NonNull Photo photo, @NonNull Photo t1) {
            return photo.getUrls().getRegular().equals(t1.getUrls().getRegular());
        }
    };
}

File: app/src/main/java/com/example/unsplash/view/adapters/MyPagedListAdapter.kt
package com.example.unsplash.view.adapters

import android.content.Context
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.paging.PagedListAdapter
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.bumptech.glide.request.RequestOptions
import com.example.unsplash.R
import com.example.unsplash.model.models.Photo
import kotlinx.android.synthetic.main.list_item.view.*

internal class MyPagedListAdapter(
    private val mContext: Context,
    private val photoClickListener: (View, Photo, Int) -> Unit
) : PagedListAdapter<Photo, MyPagedListAdapter.MyPagedViewHolder>(DIFF_CALLBACK) {

    internal inner class MyPagedViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        fun bind(i: Int) {
            itemView.apply {
                val photo = getItem(i)
                picture.transitionName = photo?.urls?.regular
                if (photo != null) {
                    Glide.with(mContext).load(photo.urls?.regular)
                        .apply(
                            RequestOptions().dontTransform() // this line
                        )
                        .thumbnail(0.1f).diskCacheStrategy(DiskCacheStrategy.ALL)
                        .into(picture)
                } else {
                    Log.d("mLog", "onBindViewHolder: photo is null")
                }
                picture.setOnClickListener {
                    if (photo != null) {
                        photoClickListener(picture, photo, i)
                    }
                }
            }
        }
    }

    override fun onCreateViewHolder(viewGroup: ViewGroup, i: Int): MyPagedViewHolder {
        val view: View = LayoutInflater.from(mContext)
            .inflate(R.layout.list_item, viewGroup, false)
        return MyPagedViewHolder(view)
    }

    override fun onBindViewHolder(myPagedViewHolder: MyPagedViewHolder, i: Int) {
        myPagedViewHolder.bind(i)
    }

    companion object {

        private val DIFF_CALLBACK = object : DiffUtil.ItemCallback<Photo>() {
            override fun areItemsTheSame(photo: Photo, t1: Photo): Boolean {
                return photo == t1
            }

            override fun areContentsTheSame(photo: Photo, t1: Photo): Boolean {
                return photo.urls?.regular == t1.urls?.regular
            }
        }
    }
}
